sentry_streams/sentry_streams/adapters/arroyo/translator.py

```diff
@@ -0,0 +1,246 @@
+            route=value.payload.route,
+            payload=self.instance.get_value(),
+        )
+
+        return routed
+
+
+class WindowedReduce(
+    ProcessingStrategy[Union[FilteredPayload, TPayload]], Generic[TPayload, TResult]
+):
+    def __init__(
+        self,
+        window_size: Union[int, float],
+        window_slide: Union[int, float],
+        acc: Callable[[], Accumulator[Any, Any]],
+        next_step: ProcessingStrategy[TResult],
+    ) -> None:
+
+        window_count = math.ceil(window_size / window_slide)
+
+        self.window_count = window_count
+        self.window_size = window_size
+        self.window_slide = window_slide
+        self.acc = acc
+        self.next_step = next_step
+
+        # initializers depending on whether it is count-based or time-based
+        if isinstance(window_size, int):
+            self.count_mode = True
+            self.message_count = 1
+
+        else:
+            self.count_mode = False
+            self.start_time = time.time()
+
+        # build the list of Reduces, as we know the number of windows to manage upfront
+        self.windows = [self.create_reduce() for _ in range(self.window_count)]
+
+        # we know the starting value of every window goes through a cycle
+        # this represents the size of that cycle / loop
+        self.window_loop = self.window_count * self.window_slide
+
+        # since we're adding elements to a cycle effectively,
+        # we know, for each window, which indices within that cycle
+        # CANNOT be part of that window
+
+        # for each window, track a 'greater' and a 'lesser' value
+        # every incoming message will be compared to these to determine
+        # whether or not the message can be added to the window
+        self.boundaries = []
+        for i in range(self.window_count):
+            greater = (self.window_size + i * self.window_slide) % self.window_loop
+
+            # this branch is because of the different meaning of window_size
+            # between count and time
+            # e.g. count window of 5 msgs: [msg1 ... msg5]
+            # e.g. time window of 5 seconds [0s ... 5s]
+            if self.count_mode:
+                less = i * self.window_slide + 1
+            else:
+                less = i * self.window_slide
+
+            self.boundaries.append((greater, less))
+
+    def create_reduce(self) -> Reduce[Any, Any]:
+        final_acc = ArroyoAccumulator(self.acc)
+
+        if self.count_mode:
+            return Reduce(
+                int(self.window_size),
+                float("inf"),
+                final_acc.accumulator,
+                final_acc.initial_value,
+                self.next_step,
+            )
+
+        else:
+            # Unfortunate
+            return Reduce(
+                sys.maxsize,
+                self.window_size,
+                final_acc.accumulator,
+                final_acc.initial_value,
+                self.next_step,
+            )
```

Comment:
Why creating multiple Reduce steps?
I think you may be better off and with a simple design if you implemented the logic to open, close and merge windows directly in this step and ignore the arroyo implementation.